### PR TITLE
Multipart support

### PR DESCRIPTION
**Pull request by egonelbre**
basic multipart upload support

Currently reuploading failed parts doesn't work.

### CODE REVIEW

File: pkg/miniogw/multipart.go [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+	upload, err := uploads.Create(bucket, object, metadata)
+	if err != nil {
+		return "", err
+	}
+
+	objectStore, err := s.storj.bs.GetObjectStore(ctx, bucket)
+	if err != nil {
+		uploads.mu.Lock()
+		delete(uploads.pending, upload.ID)
+		uploads.mu.Unlock()
```
Comment (kaloyan-raev) [inline pkg/miniogw/multipart.go]: Should we encapsulate all operations that require acquiring the lock into the MultipartUploads type? It looks error prone to me to deal with the lock outside of that type.

File: pkg/miniogw/multipart.go [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+	upload.Stream.Close()
+	// wait for completion
+	result := <-upload.Done
+	// return the final info
+	return result.Info, result.Error
+}
+
+// func (s *storjObjects) ListMultipartUploads(ctx context.Context, bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (result minio.ListMultipartsInfo, err error) {
+// 	return
+// }
```
Comment (kaloyan-raev) [inline pkg/miniogw/multipart.go]: Should the above commented code be removed if we don't need it for now?

File: pkg/miniogw/multipart.go [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+	for i, p := range list.Parts {
+		first = i
+		if partNumberMarker <= p.PartNumber {
+			break
+		}
+	}
+
+	list.Parts = list.Parts[first:]
+	if len(list.Parts) > maxParts {
+		list.NextPartNumberMarker = list.Parts[maxParts+1].PartNumber
```
Comment (kaloyan-raev) [inline pkg/miniogw/multipart.go]: I think that `+1` is wrong. On the next line we are taking `[:maxParts]`, which means from 0 to `maxParts-1` inclusively. So the next marker should be set to `maxParts`, not `maxPart+1`, shouldn't it?

File: pkg/miniogw/multipart.go [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+		list.IsTruncated = true
+	}
+
+	return list, nil
+}
+
+//
+// func (s *storjObjects) CopyObjectPart(ctx context.Context, srcBucket, srcObject, destBucket, destObject string, uploadID string, partID int, startOffset int64, length int64, srcInfo minio.ObjectInfo) (info minio.PartInfo, err error) {
+// 	return
+// }
```
Comment (kaloyan-raev) [inline pkg/miniogw/multipart.go]: Remove the above commented code too?

File: cmd/s3-benchmark/main.go [hunk 5]
```diff
@@ -28,11 +27,15 @@ func main() {
...
+	count := flag.Int("count", 50, "benchmark count")
+	duration := flag.Duration("time", 2*time.Minute, "maximum benchmark time per size")
+
+	suffix := time.Now().Format("-2006-01-02-150405")
+
+	plotname := flag.String("plot", "plot"+suffix+".svg", "plot results")
 
 	sizes := &Sizes{
-		Default: []Size{{1 << 10}, {256 << 10}, {1 << 20}, {32 << 20}, {63 << 20}},
+		Default: []Size{{1 << 10}, {256 << 10}, {1 << 20}, {32 << 20}, {64 << 20}, {256 << 20}},
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply cmd/s3-benchmark/main.go]: Ah, sure, will fix.

I've just gotten quite used to typing them that way. Shift by 10 means KB, shift by 20 means MB, shift by 30 means GB etc.

File: pkg/miniogw/multipart.go (hunk 2 above)
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: I don't feel strongly either way. I left them here more as a TODO.

File: pkg/miniogw/multipart.go (hunk 3 above)
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: It needs to return the next part number, so the +1 part is correct. However it doesn't handle the edge case properly. Good catch.

File: cmd/s3-benchmark/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply cmd/s3-benchmark/main.go]: done

File: pkg/miniogw/multipart.go [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+	"errors"
+	"io"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	minio "github.com/minio/minio/cmd"
+	"github.com/minio/minio/pkg/hash"
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: done

File: pkg/miniogw/multipart.go (hunk 1 above)
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: done

File: pkg/miniogw/multipart.go (hunk 2 above)
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: marked them more clearly as TODO, and moved them together

File: pkg/miniogw/multipart.go (hunk 4 above)
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: marked them more clearly as TODO, and moved them together

File: pkg/miniogw/multipart.go (hunk 3 above)
Comment (egonelbre) [reply pkg/miniogw/multipart.go]: Ah, yes... it should be +1. You were right.

File: pkg/miniogw/multipart.go [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+	var first int
+	for i, p := range list.Parts {
+		first = i
+		if partNumberMarker <= p.PartNumber {
+			break
+		}
+	}
+
+	list.Parts = list.Parts[first:]
+	if len(list.Parts) < maxParts {
```
Comment (kaloyan-raev) [inline pkg/miniogw/multipart.go]: Wait... why did you change:
```
if len(list.Parts) > maxParts {
```
to 
```
if len(list.Parts) < maxParts {
```
We should return a truncated result only if the remaining parts in the list are more than `maxParts`. Not the opposite.

Comment (egonelbre) [reply pkg/miniogw/multipart.go]: ah, typo... I tried the other order maxParts < len(...), but that didn't improve readability, so moved it back, but forgot the comparison.
